src/SimpleJson.ts

```diff
@@ -0,0 +1,388 @@
+							this.state === SimpleJson.Writer.State.Array);
+			} else {
+				this.Assert(this.state === SimpleJson.Writer.State.Property ||
+							this.state === SimpleJson.Writer.State.Array);
+			}
+
+			if (this.state === SimpleJson.Writer.State.Property) {
+				this.Assert(this.childCount === 0);
+			}
+
+			if (this.state === SimpleJson.Writer.State.Array || this.state === SimpleJson.Writer.State.Property) {
+				this.IncrementChildCount();
+			}
+		}
+
+		// These getters peek all the different stacks.
+
+		private get state() {
+			if (this._stateStack.length > 0) {
+				return this._stateStack[this._stateStack.length - 1].type;
+			} else {
+				return SimpleJson.Writer.State.None;
+			}
+		}
+
+		private get childCount() {
+			if (this._stateStack.length > 0) {
+				return this._stateStack[this._stateStack.length - 1].childCount;
+			} else {
+				return 0;
+			}
+		}
+
+		private get currentCollection() {
+			if (this._collectionStack.length > 0) {
+				return this._collectionStack[this._collectionStack.length - 1];
+			} else {
+				return null;
+			}
+		}
+
+		private get currentPropertyName() {
+			if (this._propertyNameStack.length > 0) {
+				return this._propertyNameStack[this._propertyNameStack.length - 1];
+			} else {
+				return null;
+			}
+		}
+
+		private IncrementChildCount() {
+			this.Assert(this._stateStack.length > 0);
+			let currEl = this._stateStack.pop()!;
+			currEl.childCount++;
+			this._stateStack.push(currEl);
+		}
+
+		private Assert(condition: boolean) {
+			if (!condition)
+				throw Error('Assert failed while writing JSON');
+		}
+
+		// This method did not exist in the original C# code. It adds
+		// the given value to the current collection (used by Write methods).
+		private _addToCurrentObject(value: number | string | boolean | null) {
+			this.Assert(this.currentCollection !== null);
+			if (this.state === SimpleJson.Writer.State.Array) {
+				(this.currentCollection as any[]).push(value);
+			} else if (this.state === SimpleJson.Writer.State.Property) {
+				this.Assert(this.currentPropertyName !== null);
+				(this.currentCollection as Record<string, any>)[this.currentPropertyName!] = value;
+				this._propertyNameStack.pop();
+			}
+		}
+
+		// In addition to `_stateStack` present in the original code,
+		// this implementation of SimpleJson use two other stacks and two
+		// temporary variables holding the current context.
+
+		// Used to keep track of the current property name being built
+		// with `WritePropertyNameStart`, `WritePropertyNameInner` and
+		// `WritePropertyNameEnd`.
+		private _currentPropertyName: string | null = null;
+
+		// Used to keep track of the current string value being built
+		// with `WriteStringStart`, `WriteStringInner` and
+		// `WriteStringEnd`.
+		private _currentString: string | null = null;
+
+		private _stateStack: SimpleJson.Writer.StateElement[] = [];
+
+		// Keep track of the current collection being built. For instance, at
```

Comment:
What do you call a "collection" in this context?

Reply:
In that particular context I mean either arrays or objects (because they can hold collections of things). Sorry if that was confusing, I'll clarify.